### ISSUE TITLE
Metal MTP head-forward numerics — alpha 0.26 vs 0.95 (Vulkan f16) / 0.96 (CPU f32)

### ISSUE THREAD

**Issue by mxaddict** [kryptic-sh/infr#39]
Follow-up from #36 (merged), tracking the Metal-specific head-numerics gap so it doesn't live only in PR prose. Related tracking precedent: #30, #34.

**Symptom**: qwen35 MTP draft acceptance on Metal is α=0.26 vs Vulkan f16's ~0.95 and the CPU exact-f32 oracle's 0.96 (Qwen3.5-4B-MTP UD-Q4_K_XL, M3 Pro). Output stays CORRECT (the spec ≡ target-only-greedy invariant holds — verify catches the bad drafts), but at α=0.26 every reject pays the qwen35 no-rewind full reprefill, so `INFR_MTP=1` is net-negative on Metal until fixed.

**What's already ruled out** (from #36's isolation):
- Not the port logic: CPU driver (same core loop) hits 0.96.
- Not generic f16 rounding: Vulkan f16 hits ~0.95.
- Not the trunk: Metal spec-decode verify is token-identical (existing pinned test).
→ The divergence is inside the HEAD forward's Metal execution: the blk.{n_layer} attention layer at hd=256 (interleaved q+gate split, QkNormRope, sigmoid out-gate) and/or the eh_proj concat path, under Metal's kernels.

**Tools in place**:
- `generate_mtp_spec_cpu_timed` = the exact-f32 acceptance oracle to A/B against (lands in #36).
- Two new Metal parity tests from #36 (`qknormrope_hd256_partial_parity`, `attention_decode_hd256_short_kv_parity`) — both green on the M3, so the bug is in an op/shape combination they don't yet cover. Suggested next probes: the eh_proj CopyStrided-concat + Linear at [1, 2ne], the per-head interleaved q/gate strided views at hd256 on Metal, and a per-op h_mtp diff (CPU vs Metal, same inputs) through one head forward — the first op whose output diverges beyond f16 noise is the culprit.

**Bar**: Metal α within noise of Vulkan's (~0.95 on the same prompts), and `INFR_MTP=1` net-positive on the M3.

**Comment by digitaloten** [comment]
Resolved — the gap was the replay tape (#38), not head numerics. Re-measured on M3 hardware against current `main`.

**Rig:** M3 Pro, macOS 26.5.2. Model: Qwen3.5-4B-MTP UD-Q4_K_XL (the same one from #36).

Metal MTP self-speculation via `INFR_METAL=1 infr bench` (the Metal accept-rate path added in #42):

```
tg128: 32.7 t/s | mtp128: 42.4 t/s (1.30x, alpha=0.95, draft 23% verify 64% catchup 12%)
tg256: 33.2 t/s | mtp256: 44.2 t/s (1.33x, alpha=0.97, draft 23% verify 64% catchup 13%)
tg128: 32.8 t/s | mtp128: 39.6 t/s (1.21x, alpha=0.95, draft 21% verify 68% catchup 11%)
```

**Metal alpha = 0.95–0.97**, matching Vulkan f16 (0.95) and the CPU f32 oracle (0.96). The 0.26 this issue tracked is gone — and it was never head numerics: it was the decode-replay tape fingerprint collision fixed in **#38** (the head recompiled a rows==1 rope+attn graph per draft step, the allocator reused a freed buffer address, the tape fingerprint collided and replayed a structurally-stale tape → intermittent all-zero logits → spurious rejects). Every op in the head was already parity-correct at 5e-3 (the hd256 short-kv attention + hd256 partial-rope tests added in #36 confirm it); nothing in the head forward needed fixing.

Consequences now that alpha is healthy:
- MTP is **net-positive on Metal (1.2–1.3×)**, where at alpha 0.26 every reject paid the qwen35 no-rewind full reprefill and it was net-negative.
- Verify is 64–68% of the cycle (the lm_head GEMV over the large vocab per draft step) — that is the honest remaining cost, and it is the dispatch-fusion workstream in **#35**, not a numerics bug.
- Output stays **token-identical to greedy** (`INFR_MTP=1` vs plain greedy, temp 0 — byte-identical 200-token generation).

Suggest closing this as resolved by #38 (+ #42 for the measurement path). Happy to leave the perf side to #35.
